Add dedicated Groq model provider

oliverbarnes opened the pull request:
Closes #273 

Features not supported by Groq:

- `json_mode: strict`
- `tool_choice: specific`
- `allowed_tools`
- streaming in json mode

Batch processing seems to be supported and documented:

https://console.groq.com/docs/batch
https://console.groq.com/docs/api-reference#batches-create

Implementation and tests adapted accordingly. E2E and nextest suite both pass. Ran openai's E2E suite to be confident changes didn't impact other provider tests.

Manually tested with a simple chat completion (eg asking the capital of Brazil, getting correct unformatted answer) and the email data extraction tutorial example:

```
[models."llama-scout-groq"]
routing = ["groq"]

[models."llama-scout-groq".providers.groq]
type = "groq"
model_name = "meta-llama/llama-4-scout-17b-16e-instruct"

[functions.my_chatbot]
type = "chat"

[functions.my_chatbot.variants.groq]
type = "chat_completion"
model = "llama-scout-groq"

[functions.extract_email]
type = "json"
user_schema = "extract_email/user_schema.json"
output_schema = "extract_email/output_schema.json"

[functions.extract_email.variants.llama_scout_email_variant]
type = "chat_completion"
model = "llama-scout-groq"
system_template = "extract_email/llama_scout_email_variant/system.minijinja"
user_template = "extract_email/llama_scout_email_variant/user.minijinja"
```

<!--
Thank you for contributing to TensorZero!

Before submitting your PR, make sure you've read **[Contributing to TensorZero](https://github.com/tensorzero/tensorzero/blob/main/CONTRIBUTING.md)**.

In particular, make sure you've run `pre-commit` and any tests relevant to your changes (including E2E tests for changes to the gateway).

By submitting this PR, unless otherwise specified, you agree to license your contributions under the **[Apache 2.0 license](https://github.com/tensorzero/tensorzero/blob/main/LICENSE)**.
-->

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Add Groq model provider with configuration, tests, and UI updates, excluding certain features.
> 
>   - **Behavior**:
>     - Adds Groq model provider, excluding support for `json_mode: strict`, `tool_choice: specific`, `allowed_tools`, and streaming in JSON mode.
>     - Updates `tensorzero.toml` with Groq model configurations.
>   - **Tests**:
>     - Adds `groq.rs` for E2E tests, including various test cases for Groq provider.
>     - Updates `mod.rs` to include Groq in the list of providers.
>   - **UI**:
>     - Updates `ModelBadge.tsx` and `ModelBadges.tsx` to include Groq provider with specific styling.
>   - **Config**:
>     - Adds `GroqProviderConfigSchema` to `models.ts` for configuration validation.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=tensorzero%2Ftensorzero&utm_source=github&utm_medium=referral)<sup> for b01c2fb88eef2a5e43d550219c6974d3384b048d. You can [customize](https://app.ellipsis.dev/tensorzero/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

virajmehta reviewed (changes requested):
hey @oliverbarnes this isn't going to work as is -- the idea with the E2E tests is that we first include whatever is needed to test the model in the `tensorzero.toml` in the `tests/e2e` folder. Then we set up the providers in the file to point to the variants we've define in the `tensorzero.toml` so that our test generation macro can generate all the e2e tests for supported features.

You'll want to first create the required models/variants in the toml, make sure the gateway starts, then add the tests in the `groq.rs` test file in order to explicitly hit those variants. Does that make sense?

oliverbarnes commented:
Ok, hopefully I got it now. Before I was thinking groq e2e would automagically use the open ai models and variants :P

(both nextest and e2e tests pass, ran clippy and fmt)

virajmehta commented:
Hey @oliverbarnes there are a lot of things wrong with this PR as is: 
- the groq e2e tests are currently not being linked as they aren't in `tests/e2e/mod.rs`. 
- groq doesn't have a model called gpt-4o-mini
- the base URL is wrong
- there is a batch API for groq but it's not well-documented and I don't think we want to support it
- AFAIK the groq API doesn't have an embeddings endpoint
- many tests that are in the `openai.rs` don't need to be replicated in groq and instead you can just use the testing macros to generate the usual tests we run against every provider. See the test files for e.g. Fireworks AI as an example.

For this and future PRs, please test your code locally before opening the PR. I understand that this might involve getting an API key to the external service you're integrating against. In groq's case this comes with a free tier so it doesn't even cost money but at the scale we're talking about for these tests the tests would cost at most a few cents to run.
I'm going to convert this one to a draft for now.

virajmehta reviewed (changes requested):
see comment

oliverbarnes commented:
Apologies @virajmehta, in hindsight this was rushed. Will do - I actually have an api key for the end to end tests and am ok spending a few credits. Just assumed wrongly that running the e2e tests would be enough to validate.

oliverbarnes commented:
Also assumed Groq simply proxied to OpenAI. Now reading up further on the docs (which I should've done before)

oliverbarnes commented:
Update: 

After too many fixes to go into here (commits try to be self explanatory), manually testing a simple chat completion (eg asking the capital of Tokyo, getting correct unformatted answer) and the email data extraction tutorial example now both work.

About batch processing, it seems well documented to me - maybe docs were recently added? 

https://console.groq.com/docs/batch
https://console.groq.com/docs/api-reference#batches-create

I'm now investigating failure of e2e tests for tool use, which I'm posting below. Tool use is supposed to work for llama scout:

https://console.groq.com/docs/tool-use

Failing e2e tests:

**test_tool_use_allowed_tools_inference_request**

```
thread 'providers::groq::test_tool_use_allowed_tools_inference_request' panicked at tensorzero-internal/tests/e2e/providers/common.rs:6230:5:
assertion `left == right` failed
  left: "brave_search"
 right: "get_humidity"
```

**test_tool_use_tool_choice_required_inference_request**

```
thread 'providers::groq::test_tool_use_tool_choice_required_inference_request' panicked at tensorzero-internal/tests/e2e/providers/common.rs:4354:5:
assertion `left == right` failed
  left: "example_function_name"
 right: "get_temperature"
```

**test_tool_use_tool_choice_required_streaming_inference_request**

```
thread 'providers::groq::test_tool_use_tool_choice_required_streaming_inference_request' panicked at tensorzero-internal/tests/e2e/providers/common.rs:4639:21:
assertion `left == right` failed
  left: "example_function_name"
 right: "get_temperature"
```

**test_tool_use_tool_choice_specific_streaming_inference_request**

```
thread 'providers::groq::test_tool_use_tool_choice_specific_streaming_inference_request' panicked at tensorzero-internal/tests/e2e/providers/common.rs:5823:27:
called `Result::unwrap()` on an `Err` value: InvalidStatusCode(502, Response { url: "http://localhost:3000/inference", status: 502, headers: {"content-type": "application/json", "x-tensorzero-gateway-version": "2025.4.7", "content-length": "470", "date": "Tue, 29 Apr 2025 16:28:03 GMT"} })
```

**test_multi_turn_parallel_tool_use_inference_request**

```
thread 'providers::groq::test_multi_turn_parallel_tool_use_inference_request' panicked at tensorzero-internal/tests/e2e/providers/common.rs:9898:5:
assertion `left == right` failed
  left: 2
 right: 1
```

GabrielBianconi commented:
Hi @oliverbarnes - we just added some automations for checking the license/CLA/etc. Could you please accept the CLA as requested by the bot above? Thanks!

oliverbarnes commented:
I have read the Contributor License Agreement (CLA) and hereby sign the CLA.

oliverbarnes commented:
Fixing clippy issues (forgot to run it with all options before reopening the PR)

virajmehta commented on tensorzero-internal/src/inference/providers/groq.rs:
```diff
@@ -0,0 +1,3513 @@
+
+        let res = request_builder
+            .body(raw_request.clone())
+            .header(reqwest::header::CONTENT_TYPE, "application/json")
```
this is duplicated

virajmehta reviewed (changes requested):
To my knowledge Groq doesn't even offer embeddings? We won't be able to merge until we have comprehensive tests for all the functionality including embeddings and batch. Feel free to remove those two, or add tests accordingly.

oliverbarnes commented:
Hey @virajmehta, all comments addressed:

- Removed wrapper
- Removed batch inferences and embedding requests
- Hardcoded the request url for chat completions

oliverbarnes commented:
This is ready for another review, in case it wasn't on the radar.

virajmehta reviewed (dismissed):
This looks good to me and tests pass locally! I'll approve this one and I appreciate all the work that went into it.

virajmehta commented:
hey Oliver I think CI is failing due to a `cargo fmt` complaint and also you'll need to add Groq (and OpenRouter, in the other PR) to the `dummy-env-file` [here](https://github.com/tensorzero/tensorzero/blob/715c2892a4d7584995a58cee939fe589f8b6dfc2/ci/dummy-env-file.env#L4). Sorry for the trouble!

oliverbarnes commented:
Updated the dummy env file, but cargo fmt doesn't change anything for me locally. Having a look at the CI now

oliverbarnes commented:
Hm, I can't see the CI history. Looks like I should have checked before pushing again :P

virajmehta reviewed (dismissed):
LGTM. Btw, in future, please avoid force-pushing to the repo and just push naturally. We squash the PRs into commits so we'd rather have the commit history in place for reviews. Thanks for the contribution!

virajmehta commented:
Hey @oliverbarnes the docker compose check is still failing. Do you mind making sure `ci/check-all-docker-compose.sh` passes locally? thanks!

oliverbarnes commented:
> Hey @oliverbarnes the docker compose check is still failing. Do you mind making sure `ci/check-all-docker-compose.sh` passes locally? thanks!

Sure thing. So far I replicated the error with the a previous version of the gateway image, and then rebuilt the image from scratch and managed to spin it up successfully with the example docker compose file.

I'm wondering if the CI just tried to spin up a previous version of the image as well (from a docker image cache?). 

Now running the full `ci/check-all-docker-compose.sh` to ensure it passes locally. Will continue tomorrow

>  please avoid force-pushing to the repo and just push naturally

Got it, will do.

virajmehta commented:
Awesome thanks! I'll try and get this merged today.

virajmehta commented on tensorzero-internal/src/inference/providers/groq.rs:
```diff
@@ -0,0 +1,2532 @@
+
+        let res = request_builder
+            .body(raw_request.clone())
+            .header(reqwest::header::CONTENT_TYPE, "application/json")
```
this is still duplicated, see line 147. This caused an e2e test failure. Can you please remove the one on line 147?

virajmehta commented:
@oliverbarnes  I ran the tests and see this in the logs:
```
2025-05-26T23:20:30.465777Z  WARN request{method=POST uri=/inference version=HTTP/1.1}:inference{otel.name="function_inference" function_name="weather_helper" variant_name="groq" episode_id="01970ee5-ca8d-7083-8a08-05682df3f7bd" inference_id="01970ee5-cb49-7bc3-a932-61e8e63c028c" episode_id="01970ee5-ca8d-7083-8a08-05682df3f7bd"}:infer_stream{function_name=weather_helper variant_name=groq otel.name="variant_inference" stream=true}:infer_model_request_stream{model_name=llama-scout-groq}:infer_stream{model_name="llama-scout-groq" otel.name="model_inference" stream=true}: tensorzero_internal::model: Skipping cache write for stream response due to error in stream: Error from groq server: Error parsing chunk. Error: missing field `choices` at line 1 column 313
Raw response: {"error":{"message":"Failed to call a function. Please adjust your prompt. See 'failed_generation' for more details.","type":"invalid_request_error","code":"tool_use_failed","failed_generation":"Tool use failed: selected tool get_temperature does not match forced tool call name self_destruct","status_code":400}}
2025-05-26T23:20:30.466323Z  WARN request{method=POST uri=/inference version=HTTP/1.1}:inference{otel.name="function_inference" function_name="weather_helper" variant_name="groq" episode_id="01970ee5-ca8d-7083-8a08-05682df3f7bd" inference_id="01970ee5-cb49-7bc3-a932-61e8e63c028c" episode_id="01970ee5-ca8d-7083-8a08-05682df3f7bd"}: tensorzero_internal::endpoints::inference: functions.Some("weather_helper").variants."groq" failed during inference: Error from groq server: Error parsing chunk. Error: missing field `choices` at line 1 column 313
```

This means groq is throwing an error rather than return the wrong tool. This is helpful info. For now we should disable these tests and add a comment in the implementation explaining the weird behavior. (In the conversion to the Groq tool call types).

oliverbarnes commented:
Thanks, that is useful (will do). I hadn't noticed those - the logs I see for `test_parallel_tool_use_streaming_inference_request` (the test that's breaking) don't have anything about the missing choices field:

```
2025-05-26T23:50:25.678637Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:25.678811Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:26.273873Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=595 ms status=200
2025-05-26T23:50:26.274009Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=595 ms status=200
2025-05-26T23:50:29.822439Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:29.822585Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:30.042259Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=219 ms status=200
2025-05-26T23:50:30.042393Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=219 ms status=200
2025-05-26T23:50:33.243961Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:33.244106Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_request: started processing request
2025-05-26T23:50:33.419869Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=175 ms status=200
2025-05-26T23:50:33.419926Z DEBUG request{method=POST uri=/inference version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=176 ms status=200
```

oliverbarnes commented:
> For now we should disable these tests and add a comment in the implementation explaining the weird behavior. (In the conversion to the Groq tool call types).

@virajmehta done. Please let me know if the comment is in the place you were thinking (or if you'd like me to go into more detail)

virajmehta commented:
@oliverbarnes thanks for the contribution! glad we got this in
